torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -3325,8 +3312,15 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   // stream, we don't need to do anything for tensor lifetime management.
   // Otherwise, we need to stage the tensors will `work.wait()`.
   if (asyncOp) {
-    work->stashTensors(inputs);
-    work->stashTensors(outputs);
+    if (coalescingState_.active) {
+      coalescingState_.tensors.insert(
+          coalescingState_.tensors.end(), inputs.begin(), inputs.end());
+      coalescingState_.tensors.insert(
+          coalescingState_.tensors.end(), outputs.begin(), outputs.end());
+    } else {
+      work->stashTensors(inputs);
+      work->stashTensors(outputs);
+    }
```

Comment:
This is the real fix to Torchrec's failure. It is using `reduce_scatter_v` and `all_gather_v`